### DISCUSSION TITLE
How to choose CUDA device if multiple are available?

### DISCUSSION THREAD

**Discussion by M-Amrollahi** [ultralytics/ultralytics#19989] [Q&A]
When I choose any device on my machine which includes 4 GPU, it still uses the ```cuda:0```. Also, I have specified the device like:
```python
model.to("cuda:2")
results = model.train(data="./result.yaml", augment=True, batch=4, imgsz=800, device="2")
```
![image](https://github.com/user-attachments/assets/316a9bd5-7891-4fb4-ac7d-2db6d6545b27)

What is the issue and how to solve it?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @M-Amrollahi, thank you for reaching out and using Ultralytics 🚀! 

We recommend checking out the [Docs](https://docs.ultralytics.com/) for guidance, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Many common questions, including device configuration, may already be addressed there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further. This should include a simple script that demonstrates the issue and any relevant configurations. 

For custom training ❓ questions, please share more context, including training logs, dataset details, and any additional code snippets. Also, ensure you're following the [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

## Upgrade

First, ensure you're using the latest version of the `ultralytics` package to rule out any resolved issues. Upgrade with:

```bash
pip install -U ultralytics
```

## Environments

YOLO can run in various verified environments with preinstalled dependencies, including [Python](https://www.python.org/), [CUDA](https://developer.nvidia.com/cuda-zone), and [PyTorch](https://pytorch.org/):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests validate YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu systems daily.

Lastly, an Ultralytics engineer will review your discussion and provide more detailed assistance soon. For real-time support, you can also join our [Discord](https://discord.com/invite/ultralytics) community 🎧.

**Comment by glenn-jocher** [comment]
@M-Amrollahi to select a specific CUDA device during training, use the `device` argument with integer indices (not strings) in your `train()` call. For your case with 4 GPUs, set `device=2` directly:

```python
results = model.train(..., device=2)
```

This will utilize `cuda:2`. For multi-GPU training, pass a list like `device=[0,2]`. You can verify available devices with `torch.cuda.device_count()` and monitor usage via `nvidia-smi`. For more details, see the device selection documentation in [Troubleshooting Common YOLO Issues](https://docs.ultralytics.com/guides/yolo-common-issues/).

**Reply by M-Amrollahi** [reply]
Actually, it is supposed to use ```cuda:1``` but it still utilize the ```cuda:0```.
```python
Ultralytics 8.3.100 🚀 Python-3.12.7 torch-2.5.1 CUDA:1 (NVIDIA GeForce GTX 1080, 8105MiB)
```
However, it uses a bit from ```cuda:1```:

![image](https://github.com/user-attachments/assets/5067aee5-a871-4996-8892-8e76e5649453)

**Reply by glenn-jocher** [reply]
@M-Amrollahi to ensure proper CUDA device selection in your training configuration, use integer indices without quotes for the `device` parameter and verify GPU visibility. Try this adjustment:

```python
results = model.train(..., device=2)  # integer instead of string
```

If issues persist, check your CUDA environment variables and GPU visibility with `torch.cuda.device_count()` before training. For advanced configuration, refer to the PyTorch device management documentation.

**Reply by M-Amrollahi** [reply]
Yes, I know. I have tried it but still does not work:
```python
import torch
import os

print(f"CUDA Available: {torch.cuda.is_available()}")
print(f"Device Count: {torch.cuda.device_count()}")
for i in range(torch.cuda.device_count()):
    print(f"Device {i}: {torch.cuda.get_device_name(i)}")
print(os.environ['CUDA_VISIBLE_DEVICES'])
```
```
CUDA Available: True
Device Count: 4
Device 0: NVIDIA GeForce GTX 1080
Device 1: NVIDIA GeForce GTX 1080
Device 2: NVIDIA GeForce GTX 1080
Device 3: NVIDIA GeForce GTX 1080
0,1,2,3
```

```python
results = model.train(data="./result.yaml", augment=True, batch=4, imgsz=800, device=1,...)
```

**Reply by glenn-jocher** [reply]
To ensure proper CUDA device selection, please check if other environment variables or scripts are overriding `CUDA_VISIBLE_DEVICES`. Try explicitly setting it before training:
```python
os.environ["CUDA_VISIBLE_DEVICES"] = "2"  # forces CUDA:0 to map to physical GPU 2
results = model.train(..., device=0)
```
This resolves device mapping conflicts when multiple GPUs are available. For full implementation details, see the [select_device documentation](https://docs.ultralytics.com/reference/utils/torch_utils/#ultralytics.utils.torch_utils.select_device).

**Comment by Y-T-G** [comment]
What's the full code you're using? It should work without `model.to()`
